lib/ide/json-rpc/src/handler.rs

```diff
@@ -150,55 +134,141 @@ pub type OngoingCalls = HashMap<Id,oneshot::Sender<ReplyMessage>>;
+        with(self.rc.borrow_mut(), |mut data| {
+            data.ongoing_calls.insert(id,sender);
+        })
+    }
+
+    /// Removes the request from the map of ongoing requests, e.g. because it
+    /// has been successfully completed or we know that we will not receive an
+    /// answer anymore.
+    ///
+    /// Returns the channel handle for the request, it should be immediately
+    /// after notified or dropped.
+    pub fn remove_ongoing_request(&self, id:Id) -> Option<oneshot::Sender<ReplyMessage>> {
+        with(self.rc.borrow_mut(), |mut data| {
+            data.ongoing_calls.remove(&id)
+        })
+    }
+
+    /// Removes all the ongoing requests. This will be recognized by the `Future`s
+    /// as losing connection error.
+    pub fn clear_ongoing_requests(&self) {
+        with(self.rc.borrow_mut(), |mut data| {
+            data.ongoing_calls.clear()
+        })
+    }
+
+    /// Obtains an id for a new request to be made.
+    pub fn generate_new_id(&self) -> Id {
+        with(self.rc.borrow_mut(), |mut data| {
+            data.id_generator.generate()
+        })
+    }
+
+    /// Sends a text message to the peer.
+    pub fn send_text_message(&self, text:String) -> std::result::Result<(), failure::Error> {
+        with(self.rc.borrow_mut(), |mut data| {
+            data.transport.send_text(text)
+        })
+    }
+
+    /// Obtains stream of events from our transport layer.
+    ///
+    /// Calling this function invalidates (closes) any previous stream obtained
+    /// from this function.
+    ///
+    /// This method is not public, as it is used internally by the `Handler` and
+    /// calling it by a third-party can break this type semantics.
+    fn transport_event_stream(&self) -> impl Stream<Item = TransportEvent> {
+        let (event_tx, event_rx) = unbounded();
+        with(self.rc.borrow_mut(), |mut data| {
+            data.transport.set_event_tx(event_tx);
+        });
+        event_rx
+    }
+
+    /// Creates a new stream with events from this handler.
+    ///
+    /// If such stream was already existing, it will be finished (and
+    /// continuations should be able to process any remaining events).
+    pub fn handler_event_stream(&self) -> impl Stream<Item = Event<Notification>> {
+        let (tx,rx)          = unbounded();
+        with(self.rc.borrow_mut(), |mut data| {
+            data.outgoing_events = Some(tx);
+        });
+        rx
+    }
+
+    /// Sends a handler event to the event stream.
+    pub fn emit_event(&self, event:Event<Notification>) {
+        with(self.rc.borrow_mut(), |data| {
+            if let Some(event_tx) = data.outgoing_events.as_ref() {
+                match event_tx.unbounded_send(event) {
+                    Ok(()) => {},
+                    Err(e) =>
+                        if e.is_full() {
+                            // Impossible, as per `futures` library docs.
+                            panic!("unbounded channel should never be full")
```

Comment:
We agreed that the errors will be proper sentences. We need to keep one style everywhere. We can talk about it once again, but if we are still on the smae page, lets make them proper sentecnes here :)